### PR TITLE
feat(streaming): onboard streaming module 

### PR DESCRIPTION
**Pull request by tyyzqmf**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary
截屏2024-06-03 17 10 35

1. If pipeline ingestion server type is KDS，then you can choose to enable streaming.
2. If the pipeline is an old version, this UI component will not be visible.
3. The old version pipeline needs to upgrade the console first, then edit the pipeline and enable this streaming module.

截屏2024-04-08 11 25 30

In the application list, users can turn on and off streaming for specified apps, and real-time streaming stack will only process the enabled app data.

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [x] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting
  - [ ] streaming ingestion
    - [ ] with Redshift Serverless
    - [ ] with provisioned Redshift

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

**Review by zxkane — CHANGES_REQUESTED** [review]
- [ ] polish UI wordings

File: frontend/src/pages/pipelines/create/steps/DataProcessing.tsx [hunk 1]
```diff
@@ -1067,6 +1071,38 @@ const DataProcessing: React.FC<DataProcessingProps> = (
...
+                <SpaceBetween direction="horizontal" size="xs">
+                  <Toggle
+                    disabled={isStreamingDisabled(update, pipelineInfo)}
+                    onChange={({ detail }) => {
+                      changeEnableStreaming(detail.checked);
+                    }}
+                    checked={pipelineInfo.enableStreaming}
+                    description={
+                      <div>
+                        <Trans i18nKey="pipeline:create.enableStreamingDesc" />
```
Comment (chenhaiyun) [inline frontend/src/pages/pipelines/create/steps/DataProcessing.tsx:1096]: Why not `defaultStr(t("pipeline:create.enableStreamingDesc"))`?

File: frontend/src/pages/projects/detail/comp/ProjectPipeline.tsx [hunk 2]
```diff
@@ -334,6 +364,40 @@ const ProjectPipeline: React.FC<ProjectPipelineProps> = (
...
+                    {
+                      text: defaultStr(t('button.delete')),
+                      id: 'rm',
+                      disabled: !selectedItems.length,
+                    },
+                    {
+                      text: defaultStr(t('button.streamEnable')),
+                      id: 'stream-enable',
+                      disabled:
+                        !selectedItems.length || !disableStreamEnableButton(),
```
Comment (chenhaiyun) [inline frontend/src/pages/projects/detail/comp/ProjectPipeline.tsx:389]: `!selectedItems.length` is redundant.

File: frontend/src/pages/projects/detail/comp/ProjectPipeline.tsx [hunk 3]
```diff
@@ -334,6 +364,40 @@ const ProjectPipeline: React.FC<ProjectPipelineProps> = (
...
+                    },
+                    {
+                      text: defaultStr(t('button.streamDisable')),
+                      id: 'stream-disable',
+                      disabled:
+                        !selectedItems.length || disableStreamEnableButton(),
```
Comment (chenhaiyun) [inline frontend/src/pages/projects/detail/comp/ProjectPipeline.tsx:395]: `!selectedItems.length` is redundant.

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @tyyzqmf on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|84|< 80|
|Duplicated lines density|:white_check_mark: OK|5|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|7|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1024)
###### _updated: 5/11/2024, 10:33:35 (UTC+0)_

**Comment by zxkane** [comment]
# SonarQube Code Analytics
## Quality Gate passed

passed

## Additional information
*The following metrics might not affect the Quality Gate status but improving them will improve your project code quality.*

## Issues
bug   [0 Bugs](http://sonarqube:9000/project/issues?id=pr-1024&resolved=false&sinceLeakPeriod=true&types=BUG)

vulnerability   [0 Vulnerabilities](http://sonarqube:9000/project/issues?id=pr-1024&resolved=false&sinceLeakPeriod=true&types=VULNERABILITY)

code_smell   [0 Code Smells](http://sonarqube:9000/project/issues?id=pr-1024&resolved=false&sinceLeakPeriod=true&types=CODE_SMELL)

## Coverage and Duplications
*No data* **Coverage**

*No data* **Duplication**

File: frontend/src/ts/utils.ts [hunk 4]
```diff
@@ -323,6 +324,46 @@ export const isReportingDisabled = (
...
+  pipelineInfo?: IExtPipeline
+) => {
+  if (!update) {
+    return (
+      !pipelineInfo?.enableRedshift ||
+      pipelineInfo?.ingestionServer?.sinkType !== SinkType.KDS
+    );
+  } else {
+    return (
+      pipelineInfo?.enableStreaming ||
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply frontend/src/ts/utils.ts:361]: This module does not support closing like reporting module. If it is enabled, it cannot be closed.

File: frontend/src/apis/application.ts [hunk 5]
```diff
@@ -43,9 +43,19 @@ const deleteApplication = async (params: { id: string; pid: string }) => {
   return result;
 };
 
+const streamApplication = async (data: {
+  id: string;
+  pid: string;
+  enable: boolean;
+}) => {
+  const result: any = await apiRequest('put', `/app/${data.id}/stream`, data);
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply frontend/src/apis/application.ts:63]: create app API:` POST /api/app`

File: frontend/src/pages/projects/detail/comp/ProjectPipeline.tsx [hunk 6]
```diff
@@ -334,6 +364,40 @@ const ProjectPipeline: React.FC<ProjectPipelineProps> = (
                 >
                   {t('button.addApplication')}
                 </Button>
+                <ButtonDropdown
+                  onItemClick={(e) => {
+                    if (e.detail.id === 'rm') {
+                      confirmDeleteApplication();
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply frontend/src/pages/projects/detail/comp/ProjectPipeline.tsx:380]: added.

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @tyyzqmf on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.60|< 80|
|Duplicated lines density|:white_check_mark: OK|5.20|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1024)
###### _updated: 6/3/2024, 07:05:03 (UTC+0)_

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :exclamation: Error
- Triggered by @tyyzqmf on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:exclamation: Error|73.50|< 80|
|Duplicated lines density|:white_check_mark: OK|5.20|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|8|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1024)
###### _updated: 6/7/2024, 04:28:18 (UTC+0)_

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @zxkane on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.70|< 80|
|Duplicated lines density|:white_check_mark: OK|5.20|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|8|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1024)
###### _updated: 6/7/2024, 05:04:50 (UTC+0)_

**Review by chenhaiyun — APPROVED** [review]

File: frontend/public/locales/zh-CN/application.json [hunk 7]
```diff
@@ -69,5 +69,10 @@
     "pipelineInfo": "数据管道信息",
     "serverEdp": "服务器端点",
     "serverDomain": "服务器域"
+  },
+  "delete": {
+    "title": "删除应用程序",
+    "tip1": "您确定要删除该应用程序 ",
```
Comment (zxkane) [inline frontend/public/locales/zh-CN/application.json:75]: Could we merge tip1 and tip2 and add placeholders for variable strings?

for example, https://github.com/awslabs/clickstream-analytics-on-aws/pull/1024/files#diff-a3cca7198752321283623781b842e9c87d263f0ba76c83514735cf73c3d6b97fR315

File: frontend/src/ts/utils.ts [hunk 8]
```diff
@@ -346,6 +348,42 @@ export const isReportingDisabled = (
...
+
+export const isStreamingVisible = (
+  update?: boolean,
+  pipelineInfo?: IExtPipeline
+) => {
+  if (update) {
+    const templateVersion = SolutionVersion.Of(
+      pipelineInfo?.templateVersion ?? ''
+    );
+    return templateVersion.greaterThan(SolutionVersion.V_1_1_5);
```
Comment (zxkane) [inline frontend/src/ts/utils.ts:382]: should be `greaterThanOrEquals(V_1_2_0)`

**Review by zxkane — CHANGES_REQUESTED** [review]

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @tyyzqmf on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.50|< 80|
|Duplicated lines density|:white_check_mark: OK|5.20|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|8|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1024)
###### _updated: 6/8/2024, 02:53:53 (UTC+0)_

**Review by zxkane — APPROVED** [review]

**Pull request merged**